Fixing Errors With Consistent Unit Tests

hutchgrant opened the pull request:
## Related Issue
Resolves #213 (+ #16 )

## Summary of Changes

1. ~~Disabled webpack plugins. Tests weren't running properly to begin with because the polyfills plugin is breaking serialization~~
2. ~~Hardcoded polyfills to where they were previously in index.html template~~
3. Utilized fs-extra where possible for asynchronous file functions, mainly in test-bed.js and in lifecycles(scaffold.js, serialize.js)
4. Modified `browser.js` to fix puppeteer launch error related to opening too many instances

Ran test locally 5 times, within the same docker container greenwood is tested with on circle ci, after reproducing issue.  

I suggest the issue was async problems with our test bed, within our lifecycles, and further the new plugin system is potentially flawed.  See serialize.js and webpack.config.common.js.

thescientist13 commented on packages/cli/test/cases/build.config.error-workspace-absolute/build.config.error-workspace-absolute.spec.js:
```diff
@@ -1,18 +1,18 @@
  *   workspace: path.join(__dirname, 'noop')
  * }
- * 
+ *
```
are these coming from your IDE by chance?  It might be good to turn that feature off (for this project anyway).  All the formatting is unfortunately adding a lot a noise to this PR.

hutchgrant replied:
Yea, but what's the point of having spaces there anyway?

thescientist13 replied:
hmm... that wasn't really the concern though.

> All the formatting is unfortunately adding a lot a noise to this PR.

hutchgrant replied:
two birds with one stone?

thescientist13 replied:
Again, that wasn't really my concern.  I will try and explain on the call instead if we have time.

thescientist13 replied:
So we didn't get to it on the call, but just to emphasize the point of focus was on the _diff_ introduced by the formatting changes, regardless of it was an addition / removal.  

In general (and yes, maybe we have been lax on formatting, maybe we should add [prettier](https://prettier.io/) to the project?) but in the face of an important PR like this (bug, large feature or refactoring) formatting changes purely for the sake of formatting changes are discouraged.  It adds "noise" which is that changes presented are not actually changes to have to worry about, yet it still counts as extra files and changes to the diff that get in the way of what is really important and what a reviewer would _want_ to read, is all.  This PR would be at least few files less and definitely many lines less if just the formatting was removed, which I think speaks for itself.

Always up for consistent formatting, and as a standalone PR that is definitely doable.  So just to politely ask please turn off any auto formatting (for this project) in your IDE so these kinds of changes don't have to keep happening.  Thanks!!! ✌️

thescientist13 commented on packages/plugin-google-analytics/test/cases/default/default.spec.js:
```diff
@@ -1,116 +1,116 @@
 
-    describe('Initialization script at the end of the <body> tag', function() {
-      let inlineScript;
+    // describe('Initialization script at the end of the <body> tag', function() {
```
instead of commenting out like this, in a test, to help with readability and review from a PR perspective, I would request using `xdescribe` (or `xit`) for disabling test code.  Less line changes / diffs and makes it clearer and cleaner that a test or suite is being ignored.

hutchgrant replied:
Ah that was it, I forgot the syntax at that moment.

hutchgrant replied:
Oh right, the problem with this is that it will fail with exit code 1 still with each of the xdescribe tests listed as "pending".

hutchgrant replied:
`describe.skip()` works though

hutchgrant commented:
The polyfills plugin was a major issue for stability to begin with. Now I don't think file functions should be synchronous, what if its a network share or something that takes time to access.  Not everyone has the resources for immediate file access at a given moment. While I agree, it could work synchronously, it's _safer_ to do file loop up and read asynchronously. Especially when we do not know the size of some of these files, or number of them that we are compiling.

thescientist13 commented:
> The polyfills plugin was a major issue for stability to begin with.

Hmm, ok.  so do we feel that narrows it down?  Is it the only issue or just a contributing factor?  I'm not as sure now after seeing this report though
https://github.com/thescientist13/greenwood-getting-started/issues/8

> Now I don't think file functions should be synchronous, what if its a network share or something that takes time to access ....

I'm not trying to make anyone's life hard to be clear, but from a build perspective, having things more synchronous could add a bit more predictability to the process as a whole, _especially_ as it comes to reading, compiling, and writing files.  

I'm sure not everything has to be synchronous but I think the things we decide to keep / make async have to be 100% guaranteed to contained / idempotent / side effect free.  I don't think doing a full audit of the build pipeline would hurt to see if there's an intelligent way to approach it.

That said, I can see the argument for making everything `async / await` but again, just have to make sure there's no chance of unpredictability.

Reliability and repeatability has to be the ultimate goal here.  Run after run after run the same input should always yield the same output.  That's what i really care about the most with this.

hutchgrant commented:
We dont have the specs from thescientist13/greenwood-getting-started#8 not enough information to determine if related.

In any event, I have demonstrated its reliable and consistant now as it stands in this PR. Once you add the plugins thats where the errors come in. So if we want to narrow it down from there further, we might find your answer.

thescientist13 commented:
> not enough information to determine if related.

well, In all fairness, the same kind of error happened in the ground zero PR.
```shell
5) Build Greenwood With: 
       Custom Webpack Plugin and Default Workspace
         Running Smoke Tests: Custom Webpack Plugin and Default Workspace
           Hello World (dummy) page
             "before each" hook for "should output a hello page directory":
     Error: ENOENT: no such file or directory, open '/root/repo/packages/cli/test/cases/build.plugins-webpack/public/hello/index.html'
```

I'm not suggesting that the plugins could have unbalanced things, but still not 💯 sure if the root cause has been there from the beginning.

hutchgrant commented:
I've readded the webpack analytics plugin and I've reverted all assosciated tests.

I noticed one issue that cropped up, that was likely due to launching multiple instances of puppeteer. I've rewritten our browser lib as a class so that we can reuse the same browser instance for every single page.  This should eliminate the puppeteer error.

Last thing to do it readd polyfills plugin to see if error is still appearing or if we've completely solved the problem of consistent unit tests.

thescientist13 commented:
Thanks, will take a look again.  There's a lot going on here, and with all the formatting changes a bit harder to focus on it specifically.

So are there two distinct bugs like we talked about?  A `fs` related bug and a serialization related bug?  Are we able to identify with any specificity the root cause of either on their own by chance?

hutchgrant commented:
It was a number of `fs` bugs with lifecycles(scaffold and serialize), one in the implementation of the polyfills plugin, and a puppeteer bug that randomly occurs due to launching of too many instances.

thescientist13 commented on package.json:
```diff
@@ -26,10 +26,13 @@
     "nyc": "^14.0.0",
     "rimraf": "^2.6.3"
+  },
+  "dependencies": {
```
I can fix this up, but this will need to go into the CLI's _package.json_.  For Lerna repos, to install dependencies per package, you just go to into the root folder of that package where _package.json_ is and just run the command, e.g
```shell
$ yarn add fs-extra
```

thescientist13 commented on packages/cli/src/lib/browser.js:
```diff
@@ -1,12 +1,28 @@
+class BrowserRunner {
+
+  constructor() {
+    this.browser = '', this.renderer = '';
```
minor, but let's put each variable declaration on its own line.  Also, I think using a `''` is misleading.  Probably just better left `undefined`  or maybe `{}`

Also, what would your thoughts be on folding renderer right in browser?  Doesn't seem all that useful keeping them apart and then you don't have to pass down instances of `browser`?  Just a thought.

hutchgrant replied:
Yea just initializing, good/bad habit.

thescientist13 replied:
good habit for sure, I think setting it as an empty string may just be a little misleading is all?  

That said, not sure what closest primitive of an instance of `browser` or `renderer` would be, so 🤷‍♂

thescientist13 replied:
maybe `null`?  idk, not a huge deal

thescientist13 commented on packages/cli/src/lifecycles/scaffold.js:
```diff
@@ -23,49 +23,61 @@ const writePageComponentsFromTemplate = async (compilation) => {
+        await fs.writeFile(path.join(target, `${filename}.js`), result);
+        resolve();
       } catch (err) {
         reject(err);
```
should we add `console.error` before all `reject`?  Seems coming out of this PR, more logging couldn't hurt (for us and users), at least until we are able to implement something like #15

hutchgrant replied:
This specific thrown error is printed back in the main [index.js](https://github.com/ProjectEvergreen/greenwood/blob/master/packages/cli/index.js#L79) already.  If you follow all thrown errors they should be printed back at that same location, if not then of course we need the error output, but from what I can tell additional outputs within a catch clause elsewhere would be redundant.

thescientist13 commented on packages/cli/src/lifecycles/scaffold.js:
```diff
@@ -23,49 +23,61 @@ const writePageComponentsFromTemplate = async (compilation) => {
     });
   };
 
+  const determineTarget = (file, context) => {
```
what is target in this case?  Do we mean output directory / path?  In NodeJS land, target may be more associated with platform, like browser, nodejs, electron, etc.

I think if what we're doing is more filesystem related, we should stick that more familiar taxonomy / language where possible.

hutchgrant replied:
Fair point, it's vague.

thescientist13 commented on test/test-bed.js:
```diff
@@ -25,44 +24,70 @@ module.exports = class TestBed {
+                  await fs.copy(targetSrc, targetPath);
+                  resolve();
+                } catch (err) {
+                  reject();
```
we should probably pass the `err` to the `reject()`

hutchgrant replied:
Yes good _catch_

thescientist13 replied:
> good  _catch_

this guy.... 🤣

thescientist13 commented on packages/cli/src/lifecycles/scaffold.js:
```diff
@@ -84,24 +96,24 @@ const writeListImportFile = async (compilation) => {
 
   if (!fs.existsSync(appDir)) {
-    await fs.mkdirSync(appDir);
+    await fs.mkdirs(appDir);
```
did we miss a `Sync` on L89?

hutchgrant commented:
This PR was in regards to stable tests which was a problem due to multiple
bugs, so multiple bugs were solved and tested at once. It's easy to say
after the fact that it should be more narrow, but part of the bug was with
the polyfills plugin so I tested against your most recent solution for it.
I argue it's causing more work to seperate it only to add it again.


On Fri., Oct. 11, 2019, 7:27 a.m. Owen Buckley, <notifications@github.com>
wrote:

> *@thescientist13* commented on this pull request.
>
> Curious, I understand why #16
> <https://github.com/ProjectEvergreen/greenwood/issues/16> was included
> here in this PR but why also #9
> <https://github.com/ProjectEvergreen/greenwood/issues/9>? It had nothing
> to do with the issue at hand again, to reference the issue regarding extra
> noise from formatting only changes, only adds extra work to the review
> process of a *bug* that requires extra attention / testing / etc.
>
> I am keen on removing it here so as to keep it isolated to its own PR and
> not conflated with the primary task at hand of solving our build problems,
> if only to really hammer the issue home (again) that trying to do too much
> in one issue is very discouraged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectEvergreen/greenwood/pull/217?email_source=notifications&email_token=AAK5A3DETN47HH3TKHBCEP3QOBPIPA5CNFSM4I4SLNI2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHVSLTQ#pullrequestreview-300623310>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAK5A3HQZB4GQ3RU7YKOEHDQOBPIPANCNFSM4I4SLNIQ>
> .
>

thescientist13 commented:
>  This PR was in regards to stable tests which was a problem due to multiple
bugs, so multiple bugs were solved and tested at once.

I understand the goal.  But what does code _not part of the bug_ in question have to do with fixing the bug in question?

>  It's easy to say after the fact that it should be more narrow, but part of the bug was with
the polyfills plugin so I tested against your most recent solution for it.

which again, does _not_ include the changes submitted as part of #212 so I'm not sure the validity of this argument


> I argue it's causing more work to seperate it only to add it again.

that's fine, I will do it myself since the value is not in the convenience to us, as has always been my point each time this topic (continues) to come up.

hutchgrant commented:
Well each bug issue has an automatic reference link to this PR since you
cited them. I don't see the value in separating, removing, and redoing it
for a different PR just for the sake of having one issue per PR.

It's possible, while not encouraged, that a single PR can tackle multiple
issues at once epecially one that was definetly contained related
issues(polyfill implementation).

On Fri., Oct. 11, 2019, 8:19 a.m. Owen Buckley, <notifications@github.com>
wrote:

> This PR was in regards to stable tests which was a problem due to multiple
> bugs, so multiple bugs were solved and tested at once.
>
> I understand the goal. But what does code *not part of the bug* in
> question have to do with fixing the bug in question?
>
> It's easy to say after the fact that it should be more narrow, but part of
> the bug was with
> the polyfills plugin so I tested against your most recent solution for it.
>
> which again, does *not* include the changes submitted as part of #212
> <https://github.com/ProjectEvergreen/greenwood/pull/212> so I'm not sure
> the validity of this argument
>
> I argue it's causing more work to seperate it only to add it again.
>
> that's fine, I will do it myself since the value is not in the convenience
> to us, as has always been my point each time this topic (continues) to come
> up.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectEvergreen/greenwood/pull/217?email_source=notifications&email_token=AAK5A3CEA5XUHMW7DZFJLK3QOBVOHA5CNFSM4I4SLNI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEA72BQI#issuecomment-541040833>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAK5A3GL33MKXYEK3F7H5HLQOBVOHANCNFSM4I4SLNIQ>
> .
>

thescientist13 commented:
I understand these nuanced discussions never seem to translate well over GitHub unfortunately but I know we've talked about them at length over calls and Slack so I'll just defer to all the same reasons I've given before for why I care about efficient PRs and git histories. 

Just try and understand beyond the surface level why I care about things like avoiding unnecessary formatting and mixing issues into single PRs.  In this case, the scope was trying to bring stability to our codebase, in the current version (0.3.x) and the WIP version for 0.4.0.  I appreciate your enthusiasm for wanting to take on all the things but please try and exercise restraint and discipline when submitting work that _others have to read, review, and test_ and be willing to work towards a commit history for the project that is clean and easy to manage and more importantly, easy to go back over.  Linking all the issues isn't the first thing you see when reviewing git history, it's the branch names, which is the whole point of putting issues in the branch name, and thus why it's my preference to keep 1 issue:1 PR so changes aren't just magically showing up in the codebase without a clean history.

Anyway, this PR is shaping up nice now, thank you for making the PR review feedback suggestions.  I'll get to doing a final review / test on those  and back out the changes to #212 so that it can stand as its own discrete change into our release branch and overall project history.

0.4.0 here we come!!

thescientist13 commented:
Also, to be clear, #9 is just an enhancement on _top_ of the polyfills plugin, to use the loader _instead_ of the single bundle file.  So everything about the polyfills plugin is already in the release branch.   We could ship without #212 and still have a polyfills plugin, it would just be less efficient at runtime.

thescientist13 commented:
@hutchgrant 
Did you have any thoughts on either of these comments I left by chance?  I think these should be considered as items for tracking.

### [_.greenwood/_ directory](https://github.com/ProjectEvergreen/greenwood/pull/217#pullrequestreview-300385867)
> One thought: Is the output to .greenwood/ anything we need to worry about between builds? Should we be blowing that out between each run? 🤔


### [browser + renderer](https://github.com/ProjectEvergreen/greenwood/pull/217#discussion_r333762162)
> Also, what would your thoughts be on folding renderer right in browser? Doesn't seem all that useful keeping them apart and then you don't have to pass down instances of browser? Just a thought.
